### ISSUE TITLE
wrong_self_convention for complex input types

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust-clippy#7219]
Lint name:wrong_self_convention

I tried this code:

```rust
trait Foo {
    fn as_byte_slice(slice: &[Self]) -> &[u8];
}
```

I expected to see this happen: No lint

Instead, this happened: 

```
warning: methods called `as_*` usually take self by reference or self by mutable reference; consider choosing a less ambiguous name
  --> utils/zerovec/src/ule/mod.rs:45:29
   |
45 |     fn as_byte_slice(slice: &[Self]) -> &[u8];
   |                             ^^^^^^^
   |
   = note: `#[warn(clippy::wrong_self_convention)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention
```

We should probably allow this when the input is _any_ reference type that involves `Self`.

### Meta

- `cargo clippy -V`: clippy 0.1.51 (2fd73fa 2021-03-23)
- `rustc -Vv`:
  ```
rustc 1.51.0 (2fd73fabe 2021-03-23)
binary: rustc
commit-hash: 2fd73fabe469357a12c2c974c140f67e7cdd76d0
commit-date: 2021-03-23
host: x86_64-unknown-linux-gnu
release: 1.51.0
LLVM version: 11.0.1
  ```

**Comment by ThibsG** [comment]
same as #7179
